### ISSUE TITLE
Error: roots not implemented

### ISSUE THREAD

**Issue by orswan** [JuliaApproximation/ApproxFun.jl#458]
A quick perusal of the open issues didn't indicate that this one is on anybody's radar:

    julia> f = Fun(((x,y) -> sin(x+y)),Interval(-1,1)^2);

    julia> abs(f)
    ERROR: roots not implemented for ApproxFun.Fun{ApproxFun.TensorSpace{Tuple{ApproxFun.Chebyshev{ApproxFun.Segment{Flo
    at64}},ApproxFun.Chebyshev{ApproxFun.Segment{Float64}}},ApproxFun.RealBasis,ApproxFun.ProductDomain{Tuple{ApproxFun.
    Segment{Float64},ApproxFun.Segment{Float64}},FixedSizeArrays.Vec{2,Float64},2},2},Float64}
     in roots(::ApproxFun.Fun{ApproxFun.TensorSpace{Tuple{ApproxFun.Chebyshev{ApproxFun.Segment{Float64}},ApproxFun.Cheb
    yshev{ApproxFun.Segment{Float64}}},ApproxFun.RealBasis,ApproxFun.ProductDomain{Tuple{ApproxFun.Segment{Float64},Appr
    oxFun.Segment{Float64}},FixedSizeArrays.Vec{2,Float64},2},2},Float64}) at C:\Users\swan\.julia\v0.5\ApproxFun\src\Ex
    tras\roots.jl:61
     in abs(::ApproxFun.Fun{ApproxFun.TensorSpace{Tuple{ApproxFun.Chebyshev{ApproxFun.Segment{Float64}},ApproxFun.Chebys
    hev{ApproxFun.Segment{Float64}}},ApproxFun.RealBasis,ApproxFun.ProductDomain{Tuple{ApproxFun.Segment{Float64},Approx
    Fun.Segment{Float64}},FixedSizeArrays.Vec{2,Float64},2},2},Float64}) at C:\Users\swan\.julia\v0.5\ApproxFun\src\Extr
    as\specialfunctions.jl:20

**Comment by dlfivefifty** [comment]
I'm not sure how to do this: in 2D, the root of a function is a continuous curve.  Then one would want to divide the square into smooth components.  But this requires function approximation on complicated domains.

See https://github.com/daanhb/FrameFun.jl for another approach that is more robust to such problems

**Comment by daanhb** [comment]
For completeness, FrameFun only does the part that is easy. It is (relatively) easy to represent a domain defined by the condition f(x,y) >= 0, for some given f. In FrameFun you can approximate functions defined on such domains. It is much, much harder to explicitly represent the curve implicitly defined by f(x,y) = 0, precisely because it may not be smooth. It could be a point, a curve with any kind of geometric singularities, a disconnected set, ...

One can implement g=abs(f) by a composite function that equals f whenever f(x,y) >= 0 and equals -f whenever f(x,y) < 0. FrameFun currently does not have that functionality anymore, but it could easily be added back in. (But a question is what you would want to do with g afterwards, as the code does not support much else besides approximating a given function on a complicated domain.)

**Comment by daanhb** [comment]
@dlfivefifty I think ApproxFun could achieve something similar, by generalizing the notion of a piecewise space. I've started exploring ideas with a generic [partition](https://github.com/daanhb/BasisFunctions.jl/blob/master/src/maps/partition.jl) in BasisFunctions. The crucial ingredient is deciding in which piece of the partition a given point lies, and that may well depend on whether $f(x,y) >=0$ or not. Piecewise intervals are a special case with a more efficient implementation.

**Comment by dlfivefifty** [comment]
@daanhb yes I believe it's possible.  But what can you do with the result other than evaluation?  As for the roots themselves, I have a feeling that it is better to encode the curves by the bivariate polynomial.  Note that given points on a zero curve it is possible to find a polynomial that vanishes using `nullspace` on a Vandermonde matrix.  I'm sure it is also possible to generate points on the zero curve from the polynomial, so there is in some sense a "transform" between the two representations.  Whether this is sufficient would depend on usages.  (The use case that comes to mine is free boundary problems.)

@orswan what is your use case?

**Comment by orswan** [comment]
@dlfivefifty Mainly, when I bumped into this issue I was just playing around, trying to familiarize myself with ApproxFun, and I was surprised to find that abs() didn't work in 2D.  The eventual use case I had in mind was plotting complex 2D Funs, or more specifically the modulus of the function.  

I had a sort of similar issue with the DifferentialEquations package recently, in trying to plot the modulus of a complex vector solution (the standard call `plot(sol)` didn't work).  I talked to somebody with that project, and he thought a potential workaround would be to allow one to pass a function as an optional argument to `plot` which would be applied to the values before plotting them.  

It doesn't matter much to me at this point, as I can just put the values of the Fun on a lattice and then plot the modulus of that array, but it could be a nice convenience if there were a "smart" way to plot 2D complex Funs.  (Maybe there is, and I just don't know about it...)  Apparently calling `plot(f)` for a Fun just plots the real part currently, which is often not what one wants.

**Comment by dlfivefifty** [comment]
If it's just for plotting, I'd use a manual grid.

**Comment by dpsanders** [comment]
You can try
https://github.com/dpsanders/IntervalConstraintProgramming.jl
for characterising inverse images of sets under functions, including zero sets and sets defined by inequalities.
